File: src/shared/runes.ts
// Lógica/derivações sobre o catálogo de runas (runeTree.ts é dado puro gerado).
import { RUNE_EDGES, RUNE_NODES, type RuneCategory, type RuneNode } from './runeTree'
import type { RuneLevel, RuneTargetPlan, RuneTargetStep } from './types'

// Categorias oficiais do mapa público (taskbarherowiki.com), com rótulos pt-BR.
export const RUNE_CATEGORY_META: Record<RuneCategory, { label: string; color: string }> = {
  chests: { label: 'Baús', color: '#d2a24c' },
  hero: { label: 'Herói', color: '#f85149' },
  gold: { label: 'Ouro', color: '#ffb020' },
  exp: { label: 'EXP', color: '#3fb950' },
  slots: { label: 'Ranhuras', color: '#a371f7' },
  offline: { label: 'Offline', color: '#56d4dd' },
  cube: { label: 'Cubo', color: '#f778ba' },
  combat: { label: 'Combate', color: '#db6d28' }
}

export const TOTAL_RUNE_NODES = RUNE_NODES.length

const byKey = new Map<number, RuneNode>(RUNE_NODES.map((n) => [n.key, n]))

export function runeByKey(key: number): RuneNode | undefined {
  return byKey.get(key)
}

export function runeColor(node: RuneNode): string {
  return RUNE_CATEGORY_META[node.category].color
}

/** Texto do efeito no nível informado (usa o valor do nível; cai pro nível 1). */
export function formatRuneEffect(node: RuneNode, level: number): string {
  if (!node.values.length) return node.effect.replace('{0}', '?')
  const idx = Math.max(0, Math.min((level || 1) - 1, node.values.length - 1))
  return node.effect.replace('{0}', String(node.values[idx]))
}

/** Custo em ouro para subir do nível atual para o próximo (0 se não há ouro/maxado). */
export function nextLevelGoldCost(node: RuneNode, currentLevel: number): number {
  if (currentLevel >= node.maxLevel) return 0
  return node.goldCost[currentLevel] ?? 0
}

/** Soma de ouro já investido até o nível atual (custos acumulados em ouro). */
export function investedGold(node: RuneNode, currentLevel: number): number {
  let sum = 0
  for (let i = 0; i < Math.min(currentLevel, node.goldCost.length); i++) sum += node.goldCost[i] ?? 0
  return sum
}

export interface RuneProgress {
  ownedNodes: number
  totalNodes: number
  maxedNodes: number
  byCategory: Record<RuneCategory, { owned: number; total: number }>
}

/** Resume o progresso a partir dos níveis observados no save. */
export function summarizeRunes(levels: RuneLevel[]): RuneProgress {
  const levelByKey = new Map(levels.map((l) => [l.key, l.level]))
  const byCategory = Object.fromEntries(
    Object.keys(RUNE_CATEGORY_META).map((c) => [c, { owned: 0, total: 0 }])
  ) as RuneProgress['byCategory']

  let ownedNodes = 0
  let maxedNodes = 0
  for (const node of RUNE_NODES) {
    const lv = levelByKey.get(node.key) ?? 0
    byCategory[node.category].total++
    if (lv > 0) {
      ownedNodes++
      byCategory[node.category].owned++
      if (lv >= node.maxLevel) maxedNodes++
    }
  }
  return { ownedNodes, totalNodes: RUNE_NODES.length, maxedNodes, byCategory }
}

// ── Runa-alvo (R3): caminho de menor custo + ouro faltante ───────────────────

// Pré-requisitos de cada nó: os nós de origem (`from`) das arestas que chegam nele.
const parentsByKey = new Map<number, number[]>()
for (const e of RUNE_EDGES) {
  const arr = parentsByKey.get(e.to)
  if (arr) arr.push(e.from)
  else parentsByKey.set(e.to, [e.from])
}

// Custo em ouro de um intervalo de níveis [from, to) de um nó (0 quando não é ouro).
// nonGold = algum nível do intervalo tem custo que não é ouro (soul stones).
function levelGold(node: RuneNode, from: number, to: number): { gold: number; nonGold: boolean } {
  let gold = 0
  let nonGold = false
  for (let i = from; i < to; i++) {
    const c = node.goldCost[i] ?? 0
    if (c > 0) gold += c
    else nonGold = true
  }
  return { gold, nonGold }
}

interface ChainResult {
  cost: number // ouro para desbloquear (nível 1) este nó e seus pré-requisitos pendentes
  chain: number[] // nós pendentes em ordem raiz→nó (já desbloqueados não entram)
  nonGold: boolean // algum passo do caminho custa não-ouro
  reachable: boolean
}

const UNREACHABLE: ChainResult = { cost: Infinity, chain: [], nonGold: false, reachable: false }

// Caminho mais barato para desbloquear `key` (nível ≥ 1) a partir de um nó já possuído
// ou da raiz, considerando seus pré-requisitos. Memoizado; guarda contra ciclos.
function cheapestUnlock(
  key: number,
  levelByKey: Map<number, number>,
  memo: Map<number, ChainResult>,
  stack: Set<number>
): ChainResult {
  if ((levelByKey.get(key) ?? 0) > 0) return { cost: 0, chain: [], nonGold: false, reachable: true }
  const cached = memo.get(key)
  if (cached) return cached
  if (stack.has(key)) return UNREACHABLE // ciclo: ignora este ramo

  const node = byKey.get(key)
  if (!node) return UNREACHABLE

  stack.add(key)
  const parents = parentsByKey.get(key) ?? []
  let bestParent: ChainResult = { cost: 0, chain: [], nonGold: false, reachable: true }
  if (parents.length > 0) {
    bestParent = UNREACHABLE
    for (const p of parents) {
      const r = cheapestUnlock(p, levelByKey, memo, stack)
      if (r.reachable && r.cost < bestParent.cost) bestParent = r
    }
  }
  stack.delete(key)

  let result: ChainResult
  if (!bestParent.reachable) {
    result = UNREACHABLE
  } else {
    const own = levelGold(node, 0, 1) // custo de desbloquear (nível 0→1)
    result = {
      cost: bestParent.cost + own.gold,
      chain: [...bestParent.chain, key],
      nonGold: bestParent.nonGold || own.nonGold,
      reachable: true
    }
  }
  memo.set(key, result)
  return result
}

/**
 * Plano de ouro até a runa-alvo (R3): soma o custo dos pré-requisitos pendentes
 * (caminho de menor custo até a raiz) + os níveis restantes do próprio alvo (até o
 * máximo), subtrai o ouro atual e devolve o progresso. Retorna null se a chave não
 * existir no catálogo.
 */
export function planRuneTarget(
  targetKey: number,
  levels: RuneLevel[],
  gold: number | null
): RuneTargetPlan | null {
  const node = byKey.get(targetKey)
  if (!node) return null

  const levelByKey = new Map(levels.map((l) => [l.key, l.level]))
  const currentLevel = levelByKey.get(targetKey) ?? 0

  // Pré-requisitos: caminho mais barato que desbloqueia todos os pais do alvo.
  const memo = new Map<number, ChainResult>()
  const parents = parentsByKey.get(targetKey) ?? []
  let prereq: ChainResult = { cost: 0, chain: [], nonGold: false, reachable: true }
  if (parents.length > 0) {
    prereq = UNREACHABLE
    for (const p of parents) {
      const r = cheapestUnlock(p, levelByKey, memo, new Set())
      if (r.reachable && r.cost < prereq.cost) prereq = r
    }
  }

  const steps: RuneTargetStep[] = []
  for (const k of prereq.chain) {
    const n = byKey.get(k)
    if (!n) continue
    const g = n.goldCost[0] ?? 0
    steps.push({
      key: k,
      name: n.name,
      icon: n.icon,
      category: n.category,
      fromLevel: 0,
      toLevel: 1,
      goldCost: g,
      payableInGold: g > 0,
      isTarget: false,
      affordable: false
    })
  }

  const alreadyComplete = currentLevel >= node.maxLevel
  const targetLeveling = levelGold(node, currentLevel, node.maxLevel)
  if (!alreadyComplete) {
    steps.push({
      key: targetKey,
      name: node.name,
      icon: node.icon,
      category: node.category,
      fromLevel: currentLevel,
      toLevel: node.maxLevel,
      goldCost: targetLeveling.gold,
      payableInGold: targetLeveling.gold > 0,
      isTarget: true,
      affordable: false
    })
  }

  const totalGoldCost = steps.reduce((sum, s) => sum + s.goldCost, 0)
  const goldHave = gold

  // Marca, respeitando a sequência, até onde o ouro atual cobre o custo acumulado.
  let running = 0
  for (const step of steps) {
    running += step.goldCost
    step.affordable = goldHave !== null && running <= goldHave
  }

  const goldMissing = goldHave === null ? totalGoldCost : Math.max(0, totalGoldCost - goldHave)
  const progress =
    totalGoldCost <= 0
      ? 1
      : goldHave === null
        ? 0
        : Math.max(0, Math.min(1, goldHave / totalGoldCost))

  return {
    targetKey,
    targetName: node.name,
    targetIcon: node.icon,
    category: node.category,
    currentLevel,
    maxLevel: node.maxLevel,
    reachable: prereq.reachable,
    alreadyComplete,
    steps,
    totalGoldCost,
    goldHave,
    goldMissing,
    progress,
    hasNonGold: prereq.nonGold || targetLeveling.nonGold
  }
}

File: dotnet/TbhTracker.Core/Logic/Runes.cs
using System.Globalization;
using TbhTracker.Core.Models;

namespace TbhTracker.Core.Logic;

// Port de src/shared/runes.ts (catalogo cru vem de runeTree.json).

public sealed class RuneCategoryMeta
{
    public string Label { get; init; } = "";
    public string Color { get; init; } = "";
}

public sealed class RuneProgressCategory
{
    public int Owned { get; set; }
    public int Total { get; set; }
}

public sealed class RuneProgress
{
    public int OwnedNodes { get; set; }
    public int TotalNodes { get; set; }
    public int MaxedNodes { get; set; }
    public Dictionary<string, RuneProgressCategory> ByCategory { get; set; } = new();
}

public static class Runes
{
    public static readonly IReadOnlyDictionary<string, RuneCategoryMeta> CategoryMeta =
        new Dictionary<string, RuneCategoryMeta>
        {
            ["chests"] = new() { Label = "Baús", Color = "#d2a24c" },
            ["hero"] = new() { Label = "Herói", Color = "#f85149" },
            ["gold"] = new() { Label = "Ouro", Color = "#ffb020" },
            ["exp"] = new() { Label = "EXP", Color = "#3fb950" },
            ["slots"] = new() { Label = "Ranhuras", Color = "#a371f7" },
            ["offline"] = new() { Label = "Offline", Color = "#56d4dd" },
            ["cube"] = new() { Label = "Cubo", Color = "#f778ba" },
            ["combat"] = new() { Label = "Combate", Color = "#db6d28" }
        };

    public static int TotalRuneNodes => Catalog.Runes.Nodes.Count;

    private static Dictionary<int, RuneNode>? _byKey;
    private static Dictionary<int, RuneNode> ByKey =>
        _byKey ??= Catalog.Runes.Nodes.ToDictionary(n => n.Key);

    private static Dictionary<int, List<int>>? _parentsByKey;
    private static Dictionary<int, List<int>> ParentsByKey
    {
        get
        {
            if (_parentsByKey != null) return _parentsByKey;
            var map = new Dictionary<int, List<int>>();
            foreach (var e in Catalog.Runes.Edges)
            {
                if (!map.TryGetValue(e.To, out var arr))
                    map[e.To] = arr = new List<int>();
                arr.Add(e.From);
            }
            return _parentsByKey = map;
        }
    }

    public static RuneNode? RuneByKey(int key) => ByKey.TryGetValue(key, out var n) ? n : null;

    public static string RuneColor(RuneNode node) =>
        CategoryMeta.TryGetValue(node.Category, out var m) ? m.Color : "#8b949e";

    public static string FormatRuneEffect(RuneNode node, int level)
    {
        if (node.Values.Count == 0) return node.Effect.Replace("{0}", "?");
        var idx = Math.Max(0, Math.Min((level <= 0 ? 1 : level) - 1, node.Values.Count - 1));
        var v = node.Values[idx];
        return node.Effect.Replace("{0}", FmtNum(v));
    }

    private static string FmtNum(double v) =>
        v == Math.Floor(v) ? ((long)v).ToString(CultureInfo.InvariantCulture)
                           : v.ToString(CultureInfo.InvariantCulture);

    public static double NextLevelGoldCost(RuneNode node, int currentLevel)
    {
        if (currentLevel >= node.MaxLevel) return 0;
        return currentLevel < node.GoldCost.Count ? node.GoldCost[currentLevel] : 0;
    }

    public static double InvestedGold(RuneNode node, int currentLevel)
    {
        double sum = 0;
        for (var i = 0; i < Math.Min(currentLevel, node.GoldCost.Count); i++) sum += node.GoldCost[i];
        return sum;
    }

    public static RuneProgress SummarizeRunes(IEnumerable<RuneLevel> levels)
    {
        var levelByKey = levels.ToDictionary(l => l.Key, l => l.Level);
        var byCategory = CategoryMeta.Keys.ToDictionary(c => c, _ => new RuneProgressCategory());

        var ownedNodes = 0;
        var maxedNodes = 0;
        foreach (var node in Catalog.Runes.Nodes)
        {
            var lv = levelByKey.TryGetValue(node.Key, out var l) ? l : 0;
            if (!byCategory.TryGetValue(node.Category, out var cat))
                byCategory[node.Category] = cat = new RuneProgressCategory();
            cat.Total++;
            if (lv > 0)
            {
                ownedNodes++;
                cat.Owned++;
                if (lv >= node.MaxLevel) maxedNodes++;
            }
        }
        return new RuneProgress
        {
            OwnedNodes = ownedNodes,
            TotalNodes = Catalog.Runes.Nodes.Count,
            MaxedNodes = maxedNodes,
            ByCategory = byCategory
        };
    }

    // ── Runa-alvo (R3) ───────────────────────────────────────────────────────

    private static (double Gold, bool NonGold) LevelGold(RuneNode node, int from, int to)
    {
        double gold = 0;
        var nonGold = false;
        for (var i = from; i < to; i++)
        {
            var c = i < node.GoldCost.Count ? node.GoldCost[i] : 0;
            if (c > 0) gold += c;
            else nonGold = true;
        }
        return (gold, nonGold);
    }

    private sealed class ChainResult
    {
        public double Cost;
        public List<int> Chain = new();
        public bool NonGold;
        public bool Reachable;
    }

    private static ChainResult Unreachable() =>
        new() { Cost = double.PositiveInfinity, Chain = new(), NonGold = false, Reachable = false };

    private static ChainResult CheapestUnlock(
        int key, Dictionary<int, int> levelByKey, Dictionary<int, ChainResult> memo, HashSet<int> stack)
    {
        if ((levelByKey.TryGetValue(key, out var lv) ? lv : 0) > 0)
            return new ChainResult { Cost = 0, Chain = new(), NonGold = false, Reachable = true };
        if (memo.TryGetValue(key, out var cached)) return cached;
        if (stack.Contains(key)) return Unreachable();

        if (!ByKey.TryGetValue(key, out var node)) return Unreachable();

        stack.Add(key);
        var parents = ParentsByKey.TryGetValue(key, out var p) ? p : new List<int>();
        var bestParent = new ChainResult { Cost = 0, Chain = new(), NonGold = false, Reachable = true };
        if (parents.Count > 0)
        {
            bestParent = Unreachable();
            foreach (var par in parents)
            {
                var r = CheapestUnlock(par, levelByKey, memo, stack);
                if (r.Reachable && r.Cost < bestParent.Cost) bestParent = r;
            }
        }
        stack.Remove(key);

        ChainResult result;
        if (!bestParent.Reachable)
        {
            result = Unreachable();
        }
        else
        {
            var own = LevelGold(node, 0, 1);
            result = new ChainResult
            {
                Cost = bestParent.Cost + own.Gold,
                Chain = new List<int>(bestParent.Chain) { key },
                NonGold = bestParent.NonGold || own.NonGold,
                Reachable = true
            };
        }
        memo[key] = result;
        return result;
    }

    public static RuneTargetPlan? PlanRuneTarget(int targetKey, IEnumerable<RuneLevel> levels, double? gold)
    {
        if (!ByKey.TryGetValue(targetKey, out var node)) return null;

        var levelByKey = levels.ToDictionary(l => l.Key, l => l.Level);
        var currentLevel = levelByKey.TryGetValue(targetKey, out var cl) ? cl : 0;

        var memo = new Dictionary<int, ChainResult>();
        var parents = ParentsByKey.TryGetValue(targetKey, out var p) ? p : new List<int>();
        var prereq = new ChainResult { Cost = 0, Chain = new(), NonGold = false, Reachable = true };
        if (parents.Count > 0)
        {
            prereq = Unreachable();
            foreach (var par in parents)
            {
                var r = CheapestUnlock(par, levelByKey, memo, new HashSet<int>());
                if (r.Reachable && r.Cost < prereq.Cost) prereq = r;
            }
        }

        var steps = new List<RuneTargetStep>();
        foreach (var k in prereq.Chain)
        {
            if (!ByKey.TryGetValue(k, out var n)) continue;
            var g = n.GoldCost.Count > 0 ? n.GoldCost[0] : 0;
            steps.Add(new RuneTargetStep
            {
                Key = k,
                Name = n.Name,
                Icon = n.Icon,
                Category = n.Category,
                FromLevel = 0,
                ToLevel = 1,
                GoldCost = g,
                PayableInGold = g > 0,
                IsTarget = false,
                Affordable = false
            });
        }

        var alreadyComplete = currentLevel >= node.MaxLevel;
        var targetLeveling = LevelGold(node, currentLevel, node.MaxLevel);
        if (!alreadyComplete)
        {
            for (int i = currentLevel; i < node.MaxLevel; i++)
            {
                var targetStep = LevelGold(node, i, i+1);

                steps.Add(new RuneTargetStep
                {
                    Key = targetKey,
                    Name = node.Name,
                    Icon = node.Icon,
                    Category = node.Category,
                    FromLevel = i,
                    ToLevel = i+1,
                    GoldCost = targetStep.Gold,
                    PayableInGold = targetStep.Gold > 0,
                    IsTarget = true,
                    Affordable = false
                });
            }
        }

        var totalGoldCost = steps.Sum(s => s.GoldCost);
        var goldHave = gold;

        double running = 0;
        foreach (var step in steps)
        {
            running += step.GoldCost;
            step.Affordable = goldHave != null && running <= goldHave;
        }

        var goldMissing = goldHave == null ? totalGoldCost : Math.Max(0, totalGoldCost - goldHave.Value);
        var progress = totalGoldCost <= 0
            ? 1
            : goldHave == null
                ? 0
                : Math.Max(0, Math.Min(1, goldHave.Value / totalGoldCost));

        return new RuneTargetPlan
        {
            TargetKey = targetKey,
            TargetName = node.Name,
            TargetIcon = node.Icon,
            Category = node.Category,
            CurrentLevel = currentLevel,
            MaxLevel = node.MaxLevel,
            Reachable = prereq.Reachable,
            AlreadyComplete = alreadyComplete,
            Steps = steps,
            TotalGoldCost = totalGoldCost,
            GoldHave = goldHave,
            GoldMissing = goldMissing,
            Progress = progress,
            HasNonGold = prereq.NonGold || targetLeveling.NonGold
        };
    }
}
